test/0.8.9/utils/pausable-util.test.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.4.24 <0.9.0;
+
+import "forge-std/Test.sol";
+import { PausableUntil } from "contracts/0.8.9/utils/PausableUntil.sol";
+
+contract ExposedPausableUntil is PausableUntil {
+    function setPausedState(uint256 _resumeSince) public {
+        _setPausedState(_resumeSince);
+    }
+
+    function pauseUntil(uint256 _pauseUntilInclusive) public {
+        _pauseUntil(_pauseUntilInclusive);
+    }
+
+    function pauseFor(uint256 _duration) public {
+        _pauseFor(_duration);
+    }
+
+    function resume() public {
+        _resume();
+    }
+
+    function checkResumed() public view {
+        _checkResumed();
+    }
+
+    function checkPaused() public view {
+        _checkPaused();
+    }
+}
+
+contract PausableUntilTest is Test {
+    ExposedPausableUntil public pausableUntil;
+
+    function setUp() public {
+        pausableUntil = new ExposedPausableUntil();
+    }
+
+    function testIsPaused() public {
+        assertEq(pausableUntil.isPaused(), false);
+    }
+
+    function testSetPausedState(uint256 _resumeSince, uint256 _random_time) public {
+        // ignore obvious 1 state and underflow condition
+        vm.assume(_resumeSince != 1);
+        vm.assume(_resumeSince >= block.timestamp);
+
+        // make sure we're not already paused
+        assertEq(pausableUntil.isPaused(), false);
+
+        // pause until _resumeSince
+        pausableUntil.setPausedState(_resumeSince);
```

Comment:
Should we check events via [expectEmit](https://book.getfoundry.sh/cheatcodes/expect-emit), wdyt?

Reply:
I looked into this, but I'm having trouble trying to sort out the syntax of expectEmit.  I see the docs suggest a true, true, true, true pattern, but it's not clear to me what that's trying to acheive.  It would seem natural to me that I should be able to say vm.expectEmit(NAME_OF_EVENT) and that seems to not work.  I tried including in the Exposed contract, but got a name space collision on defining the events because they collide with PausableUntil.   I tried including them in PausableUntilTest, but forge bombs out on expectEmit not found.  Any advise you might be able to offer on how to syntactically do this, I'm happy to implement it.

Reply:
Thank you so much for your time!

Let me try to help. If you look at the documentation, you can see that the entire event test flow is divided into 3 steps:

1. Call cheat code, like `vm.expectEmit(true, true, true, true);`  - func has 4 parameters:
 - checkTopic1 - do we need to check the first indexed topic from event
 - checkTopic2 - check the second
 - checkTopic3 - check the third
 - checkData - do we need to check/compare data in topics

2. Emit expected event with expected data. Here we "assign" expected event to foundry.

3. Call the real function that we want to test. Like `pausableUntil.setPausedState(_resumeSince);`.
 - If `checkData` was checked as true - foundry compare expected data from step2 and actual from step3.

So if briefly recap: 
 1. Prepare expected data ( your `vm.expectEmit(NAME_OF_EVENT)` divided into 2 steps : `vm.expectedEmit(...)+ emit Event(..)` )
 2. Call actual func from smart contract with event

Reply:
It would be great to have together with a special case check for the `PAUSE_INFINITELY` value here